man/org.freedesktop.systemd1.xml

```diff
@@ -2548,6 +2557,7 @@ node /org/freedesktop/systemd1/unit/avahi_2ddaemon_2eservice {
                  in  b read_only,
                  in  b mkdir,
                  in  a(ss) options);
+      DumpFileDescriptorStore(out a(suuutuusu) entries);
```

Comment:
Shouldn't we do this via a memfd containing JSON? At least that way if we ever need to add something in the future we can do so without adding a new method call.

Reply:
We can certainly add that one day if needed. But note that the dump code that uses memfd uses that for size reasons (you cannot pass arbitrarily large dumps via dbus native proto), not for a reaspm pf "we want something less rigid than the dbus type system". The size thing is a hard issue, we ran into this IRL, so we added the memfd as workaround, and it's not more than that because inherently incompatible with remoting operation via ssh or such. If we really run into issues with the message sizes for the fdstore too, then we can go down the memfd route too, as an alternative. But I'd be really careful with the other issue: yeah dbus type system is inflexible, but as long as we do use dbus we should buy into it, otherwise we degrade dbus to just a noise envelope on AF_UNIX, which brings no benefits over just using AF_UNIX directly....

I am not too concerned that we want a lot of extensibility here. fstat() hasn't changed in 40 years after all, and statx() doesn't bring much relevant to the table for this case here either. /proc/self/fd is pretty stable too. Hence I wouldn't know what to add here even if we wanted to.

Here's another way to see this: we probably should extend the existing Dump() calls to also show the fdstore – we currently don't and that really should be addressed. Dump() already does memfds, and is not bound by the dbus type system.

Hence, let's keep DumpFileDescriptorStore() separate from that. Let's have the structured, client-friendly, high-level DumpFileDescriptorStore() and then (in a separate PR) also extend the messier, low-level, free-form, memfd-based regular Dump() to show the fdstore, too.